File: simple-stack-example-mvp/src/main/java/com/zhuinden/simplestackdemoexamplemvp/presentation/paths/tasks/TasksPresenter.java
package com.zhuinden.simplestackdemoexamplemvp.presentation.paths.tasks;

import android.content.res.Resources;
import android.support.annotation.Nullable;

import com.jakewharton.rxrelay.BehaviorRelay;
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository;
import com.zhuinden.simplestackdemoexamplemvp.presentation.objects.Task;
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.addoredittask.AddOrEditTaskKey;
import com.zhuinden.simplestackdemoexamplemvp.presentation.paths.taskdetail.TaskDetailKey;
import com.zhuinden.simplestackdemoexamplemvp.util.BasePresenter;
import com.zhuinden.simplestackdemoexamplemvp.util.MessageQueue;
import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.statebundle.StateBundle;

import javax.inject.Inject;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by Owner on 2017. 01. 27..
 */
// UNSCOPED!
public class TasksPresenter
        extends BasePresenter<TasksView, TasksPresenter>
        implements Bundleable {
    @Inject
    public TasksPresenter() {
    }

    @Inject
    Backstack backstack;

    @Inject
    TaskRepository taskRepository;

    @Inject
    MessageQueue messageQueue;

    @Inject
    Resources resources;

    BehaviorRelay<TasksFilterType> filterType = BehaviorRelay.create(TasksFilterType.ALL_TASKS);

    Subscription subscription;

    @Override
    public void onAttach(TasksView view) {
        subscription = filterType.asObservable() //
                .doOnNext(tasksFilterType -> view.setFilterLabelText(tasksFilterType.getFilterText())) //
                .switchMap((tasksFilterType -> tasksFilterType.filterTask(taskRepository))) //
                .observeOn(Schedulers.computation())
                .map(tasks -> view.calculateDiff(tasks))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(pairOfDiffResultAndTasks -> {
                    if(view != null) {
                        view.showTasks(pairOfDiffResultAndTasks, filterType.getValue());
                    }
                });

        messageQueue.requestMessages(Backstack.getKey(view.getContext()), view);
    }

    @Override
    public void onDetach(TasksView view) {
        subscription.unsubscribe();
    }

    public void openAddNewTask() {
        backstack.goTo(AddOrEditTaskKey.create(Backstack.getKey(getView().getContext())));
    }

    public void openTaskDetails(Task task) {
        backstack.goTo(TaskDetailKey.create(task.id()));
    }

    public void completeTask(Task task) {
        taskRepository.setTaskCompleted(task);
        getView().showTaskMarkedComplete();
    }

    public void uncompleteTask(Task task) {
        taskRepository.setTaskActive(task);
        getView().showTaskMarkedActive();
    }

    public void deleteCompletedTasks() {
        taskRepository.deleteCompletedTasks();
        getView().showCompletedTasksCleared();
    }

    public void setFiltering(TasksFilterType filterType) {
        this.filterType.call(filterType);
    }

    @Override
    public StateBundle toBundle() {
        StateBundle bundle = new StateBundle();
        bundle.putString("FILTERING", filterType.getValue().name());
        return bundle;
    }

    @Override
    public void fromBundle(@Nullable StateBundle bundle) {
        if(bundle != null) {
            filterType.call(TasksFilterType.valueOf(bundle.getString("FILTERING")));
        }
    }
}

File: samples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/features/tasks/TasksPresenter.kt
package com.zhuinden.simplestackdemoexamplemvp.features.tasks

import androidx.recyclerview.widget.DiffUtil
import com.jakewharton.rxrelay2.BehaviorRelay
import com.zhuinden.simplestack.Backstack
import com.zhuinden.simplestack.Bundleable
import com.zhuinden.simplestackdemoexamplemvp.core.mvp.BasePresenter
import com.zhuinden.simplestackdemoexamplemvp.core.navigation.ViewKey
import com.zhuinden.simplestackdemoexamplemvp.core.navigation.getKey
import com.zhuinden.simplestackdemoexamplemvp.data.repository.TaskRepository
import com.zhuinden.simplestackdemoexamplemvp.domain.Task
import com.zhuinden.simplestackdemoexamplemvp.features.addoredittask.AddOrEditTaskKey
import com.zhuinden.simplestackdemoexamplemvp.features.taskdetail.TaskDetailKey
import com.zhuinden.simplestackdemoexamplemvp.util.BackstackHolder
import com.zhuinden.simplestackdemoexamplemvp.util.MessageQueue
import com.zhuinden.statebundle.StateBundle
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit
import javax.inject.Inject

/**
 * Created by Owner on 2017. 01. 27..
 */
// UNSCOPED!
class TasksPresenter @Inject constructor(
    private val backstackHolder: BackstackHolder,
    private val taskRepository: TaskRepository,
    private val messageQueue: MessageQueue
) : BasePresenter<TasksView>(), TasksView.Presenter, Bundleable {
    private val filterType: BehaviorRelay<TasksFilterType> = BehaviorRelay.createDefault(TasksFilterType.AllTasks())

    lateinit var disposable: Disposable

    override fun onAttach(view: TasksView) {
        super.onAttach(view)
        disposable = filterType
            .doOnNext { tasksFilterType -> view.setFilterLabelText(tasksFilterType.filterText) } //
            .switchMap { tasksFilterType -> tasksFilterType.filterTask(taskRepository) } //
            .observeOn(Schedulers.computation())
            .map<Pair<DiffUtil.DiffResult, List<Task>>> { tasks -> view.calculateDiff(tasks) }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { pairOfDiffResultAndTasks ->
                view.showTasks(pairOfDiffResultAndTasks, filterType.value!!)
            }
        messageQueue.requestMessages(Backstack.getKey<ViewKey>(view.context), view)
    }

    override fun onDetach(view: TasksView) {
        disposable.dispose()
    }

    override fun onTaskCheckClicked(task: Task) {
        if (!task.isCompleted) {
            completeTask(task)
        } else {
            uncompleteTask(task)
        }
    }

    override fun onNoTasksAddButtonClicked() {
        openAddNewTask()
    }

    override fun onTaskRowClicked(task: Task) {
        backstackHolder.backstack.goTo(TaskDetailKey(task.id))
    }

    override fun onFilterActiveSelected() {
        setFiltering(TasksFilterType.ActiveTasks())
    }

    override fun onFilterCompletedSelected() {
        setFiltering(TasksFilterType.CompletedTasks())
    }

    override fun onFilterAllSelected() {
        setFiltering(TasksFilterType.AllTasks())
    }

    override fun onClearCompletedClicked() {
        deleteCompletedTasks()
    }

    override fun onRefreshClicked() {
        view?.isRefreshing = true
        Single.just("")
            .delay(2500, TimeUnit.MILLISECONDS)
            .subscribeOn(AndroidSchedulers.mainThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { ignored -> view?.isRefreshing = false }
    }

    private fun openAddNewTask() {
        backstackHolder.backstack.goTo(AddOrEditTaskKey.AddTaskKey(view!!.getKey()))
    }

    private fun completeTask(task: Task) {
        taskRepository.setTaskCompleted(task)
        view!!.showTaskMarkedComplete()
    }

    private fun uncompleteTask(task: Task) {
        taskRepository.setTaskActive(task)
        view!!.showTaskMarkedActive()
    }

    private fun deleteCompletedTasks() {
        taskRepository.deleteCompletedTasks()
        view!!.showCompletedTasksCleared()
    }

    private fun setFiltering(filterType: TasksFilterType) {
        this.filterType.accept(filterType)
    }

    override fun toBundle(): StateBundle = StateBundle().apply {
        putParcelable("FILTERING", filterType.value)
    }

    override fun fromBundle(bundle: StateBundle?) {
        bundle?.run {
            filterType.accept(getParcelable("FILTERING")!!)
        }
    }
}
